sbncode/BeamSpillInfoRetriever/ICARUSBNBRetriever/ICARUSBNBRetriever_module.cc

```diff
@@ -0,0 +1,475 @@
+/** ********************************************************************
+ * @file ICARUSBNBRetriever_module.cc
+ * @date Wed April 9 2021
+ * @author J. Zennamo (FNAL)
+ * 
+ * Based heavily on code by Z. Pavlovic written for MicroBooNE 
+ * Based heavily on code by NOvA collaboration (Thanks NOvA!):
+ *        https://cdcvs.fnal.gov/redmine/projects/novaart/repository/entry/trunk/IFDBSpillInfo/BNBInfo_module.cc
+ * Database implementation by Justin Mueller 
+ */
+
+#include "art/Framework/Core/EDProducer.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "sbnobj/Common/POTAccounting/BNBSpillInfo.h"
+#include "sbncode/BeamSpillInfoRetriever/POTTools.h"
+#include "sbnobj/Common/Trigger/ExtraTriggerInfo.h"
+#include "sbndaq-artdaq-core/Overlays/ICARUS/ICARUSTriggerV3Fragment.hh"
+#include <sqlite3.h>
+
+namespace sbn {
+  class ICARUSBNBRetriever;
+}
+
+class sbn::ICARUSBNBRetriever : public art::EDProducer {
+public:
+  
+  struct Config {
+    
+    using Name = fhicl::Name;
+    using Comment = fhicl::Comment;
+    
+    fhicl::Atom<double> TimePadding {
+      Name{ "TimePadding" },
+      Comment{ "extension to the time window considered when collecting spills [seconds]" },
+      0.0333 // default
+      };
+ 
+    fhicl::Atom<double> BESOffset {
+      Name{ "BESOffset" },
+      Comment{ "Offset between beam early warning signal and $1D [seconds]" },
+      0.036 // default
+      };
+   
+    fhicl::Atom<std::string> RawDataLabel {
+      Name{ "raw_data_label" },
+      Comment{ "art data product instance name for trigger information (product label is 'daq')" }
+      };
+    
+    fhicl::Atom<std::string> DeviceUsedForTiming {
+      Name{ "DeviceUsedForTiming" },
+      Comment{ "name in the IFBeam database of the device used to extract spill times" }
+      };
+    
+    fhicl::Atom<std::string> URL {
+      Name{ "URL" },
+      Comment{ "IFBeam database access URL" }
+      };
+    
+    fhicl::Atom<std::string> Bundle {
+      Name{ "Bundle" },
+      Comment{ "" } // explain what this is and which database/table it's looking for
+      };
+    
+    fhicl::Atom<double> TimeWindow {
+      Name{ "TimeWindow" },
+      Comment{ "" } // explain what this is, what's for and its unit
+      };
+    
+    fhicl::Atom<std::string> MultiWireBundle {
+      Name{ "MultiWireBundle" },
+      Comment{ "" } // explain what this is and which database/table it's looking for
+      };
+    
+    fhicl::Atom<double> MWR_TimeWindow {
+      Name{ "MWR_TimeWindow" },
+      Comment{ "" } // explain what this is, what's for and its unit
+      };
+
+    fhicl::Atom<std::string> TriggerDatabaseFile {
+      Name{ "TriggerDatabaseFile" },
+      Comment{ "" } // explain what this is, what's for and its unit
```

Comment:
Could you add this:
```suggestion
      Comment{ "path of local database of all recorded events and their trigger, in SQLite format" }
```
Basically ICARUS is building an independent database of all the events and their full trigger information, to overcome the issue of the fragmentation of the information on contiguous events into many _art_/ROOT files.